Support for screenshots API

kartsims opened the pull request:
As discussed in scottgonzalez/node-browserstack#25 here is the support for screenshots API. 

I am also thinking about adding a nice callback (or option, not decided yet) to the getScreenshot method so that it waits until the job is finished to execute the callback specified. That would go a little beyond plain API support but would be a nice feature, what do you think ? We may talk about it in another issue though.

scottgonzalez commented:
Thanks! Overall this looks really good and it's great that you wrote docs with the code :-)

I'm going to go through and comment on some small issues. Let me know if you're interested in making these tweaks or if you'd rather have me do it.

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,137 @@
+var https = require("https"),
```
Missing spaces inside the parens. I know this is copy/pasted from the existing file, but I just noticed it.

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,137 @@
+	}
+
+	return a;
+}
```
Let's move `getUA()` and `extend()` into a shared module to keep the code DRY.

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,137 @@
+
+		req.end();
+	}
+});
```
I feel like we can probably remove a lot of the duplication in the request logic, but for now I'm fine with this.

scottgonzalez commented on readme.md:
```diff
@@ -41,116 +54,157 @@ A common pattern in the API is a "browser object" which is just a plain object w
+* `username`: The username for the BrowserStack account.
+* `password`: The password for the BrowserStack account.
+* `version` (optional; default: `4`): Which version of the BrowserStack API to use.
+* `server` (optional; default: `{ host: "api.browserstack.com", port: 80 }`): An object containing `host` and `port` to connect to a different BrowserStack API compatible service.
```
Why were these all outdented? It now looks like the method takes many parameters, as opposed to defining what properties exist for the `settings` parameter.

kartsims commented:
I'll take care of those, hopefully today. I'll look into the DRY stuff

About the last bit, it is a mistake from my end (IDE?) I'll check on it

kartsims commented:
I have done some refactoring work and changed a few things. Lots of code diff to read, but key points are here :

- Moved all request handling in a client.js file, browserstack.js is now just an entry point to instantiate the API clients
- Removed internal/external apis as these extended the Client.prototype, and now that the base client is shared between REST API and Screenshots API the two of them couldn't get along anymore if we extend the prototype. Consequence : methods and properties are public now, I don't think that is an issue but correct me if I am wrong
- `extend` function now uses Node's native util._extend function (we probably want to change the node supported version here then, but I couldn't find the information)
- removed this.server.port, it was not used anywhere so I just got rid of it

scottgonzalez commented:
> Moved all request handling in a client.js file, browserstack.js is now just an entry point to instantiate the API clients

That's fine as it gets rid of a bunch of duplication.

> Removed internal/external apis as these extended the Client.prototype, and now that the base client is shared between REST API and Screenshots API the two of them couldn't get along anymore if we extend the prototype.

Sure they can. You just need to properly use inheritance. There should be a base client, then an API client and a screenshot client which both inherit from the base client.

> Consequence : methods and properties are public now, I don't think that is an issue but correct me if I am wrong

These were always public, so that's not an issue. The problem is that you've gotten rid of all the prototype logic. I really dislike extending instances during construction the way that this is currently handled.

> extend function now uses Node's native util._extend function (we probably want to change the node supported version here then, but I couldn't find the information)

The underscore in that method name means that it's not public and you should not use it.

> removed this.server.port, it was not used anywhere so I just got rid of it

It's definitely used: https://github.com/scottgonzalez/node-browserstack/blob/1246998566c7d577f01b43dc56710ed1f9a198ed/lib/browserstack.js#L165

See https://github.com/scottgonzalez/node-browserstack/pull/5 for history.

kartsims commented:
With this little exercise I struggled a bit and learnt a few things about classes in JS so that is really cool. I just hope I did well so please let me know if you see anything I may improve, I really want to make the best way possible.

Anyway, I fixed the issues you pointed out.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,275 @@
+      throw new Error( "Invalid version" );
+    }
+    apiClient.prototype.version = version;
+    extend( apiClient.prototype, versionSettings);
```
This is a destructive operation. Every time a new client is created, you're changing the base prototype. This means that two clients for different versions will step on top of each other and whichever is created first will break. Please don't change the way that the versioned APIs are handled.

scottgonzalez commented:
Please don't switch from tabs to spaces for indentation.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,275 @@
+
+
+// client
+function apiClient( settings ) {
```
Constructors should start with a capital letter.

scottgonzalez commented on package.json:
```diff
@@ -11,7 +11,9 @@
   },
-  "dependencies": {},
+  "dependencies": {
+    "extend": "^3.0.0"
```
I'd prefer keeping this dependency free since we only need a very basic `extend()` implementation.

kartsims commented:
- The `extend` dependency has been removed and its basic function moved to the base client class
- Indentation fixed
- Capital letters added to constructor
- Version handling has been copy-pasted back as it were, except for the class names : Version was not very explicit so I couldn't use `instanceof` in my application. As some methods are now the same in the two classes I could implement a generic call for `getBrowsers` method and the class name is the best way to identify which client I am working with so `instanceof`comes handy there
About extend : I originally implemented it as a constructor method rather than the prototype's. But then I thought it might not belong there but rather in the prototype so that it may be used by the internal methods using the syntax `this.extend`. I am not sure of that choice, could you please tell me which is the right way to go ? You may look at the last commit to see the differences.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,281 @@
+var BaseClient = require( "./client" ),
+util = require( "util" );
```
Missing indentation.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,281 @@
+
+BaseClient.prototype.extend( ApiBaseClient.prototype, {
+
+	// public api
```
This comment belongs above the extend call.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,281 @@
+	},
+
+
+	// internal stuff
```
Please follow the convention that already existed of doing a separate extend to more clearly delineate where the internal code starts. Also, please change "stuff" back to "API".

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,281 @@
+			}
+
+			if ( parseFloat( version ) >
+				(parseFloat( latest[ browserId ] ) || 0) ) {
```
Why did the indentation change?

kartsims replied:
My IDE automatically reindented everything using tabs, I'll check on the broken `if` line

scottgonzalez commented on lib/browserstack.js:
```diff
@@ -1,367 +1,7 @@
-});
-
-
+var browserstackAPI = require( "./api" ),
```
`browserstackApi`

scottgonzalez commented on lib/client.js:
```diff
@@ -0,0 +1,97 @@
+	new Buffer( this.username + ":" + this.password ).toString( "base64" );
+};
+
+BaseClient.prototype.extend = function( a, b ) {
```
This doesn't belong in the client. It should be in its own module.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,281 @@
+util = require( "util" );
+
+
+// client
```
This comment doesn't provide any value as it simply restates what the function name tells you. Please remove.

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,50 @@
+	util = require( "util" );
+
+
+// client
```
Please remove.

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,50 @@
+
+BaseClient.prototype.extend( ScreenshotClient.prototype, {
+
+	// public API
```
This comment belongs above the `extend()` call.

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,50 @@
+	},
+
+
+	// internal stuff
```
Please move this into a separate `extend()` call. Change "stuff" to "API".

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,50 @@
+	server: {
+		host: "www.browserstack.com"
+	},
+	path: function( path ) {
```
Missing blank line above function.

scottgonzalez commented:
> Version handling has been copy-pasted back as it were, except for the class names : Version was not very explicit so I couldn't use instanceof in my application.

I'm fine with the change you made, but I'm not sure what you were trying to check on or how this made any difference since you don't have access to the constructor name either way.

> As some methods are now the same in the two classes I could implement a generic call for getBrowsers method and the class name is the best way to identify which client I am working with so instanceof comes handy there.

I'm not sure what you mean by this. Why do you care what class you're working with?

> About extend : I originally implemented it as a constructor method rather than the prototype's. But then I thought it might not belong there but rather in the prototype so that it may be used by the internal methods using the syntax this.extend. I am not sure of that choice, could you please tell me which is the right way to go ? You may look at the last commit to see the differences.

It should just be a standalone function. It's a generic helper that doesn't have anything to do with the client and users won't need to access it from the client.

kartsims commented:
OK thanks for the comments on `extend`, I was obviously on a wrong path

About the class names forget it, I was (again) wrong about accessing this property. I got confused because when I do a `console.log` on a client instance created using `BrowserStack.createClient` I could see the class name but it seems I can't access using `instanceof` or `Object.getPrototypeOf`

scottgonzalez commented:
If you can provide a scenario where you want to know this information, I can probably tell you how to accomplish what you want.

kartsims commented:
Never mind, the scenario is not finalized yet I haven't got back into this
feature yet but thanks for the offer, I'll let you know if I get stuck on
this

2015-11-18 15:58 GMT+01:00 Scott González <notifications@github.com>:

> If you can provide a scenario where you want to know this information, I
> can probably tell you how to accomplish what you want.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scottgonzalez/node-browserstack/pull/33#issuecomment-157740003>
> .
>

kartsims commented:
Anything preventing this PR from being merged ?

scottgonzalez commented:
Sorry, I've been very busy this past week. I should be able to get back to this soon and do another review.

kartsims commented:
Cool, thanks!

scottgonzalez commented on lib/client.js:
```diff
@@ -0,0 +1,89 @@
+	extend( this, settings );
+
+	this.authHeader = "Basic " +
+	new Buffer( this.username + ":" + this.password ).toString( "base64" );
```
Missing one level of indentation on this line.

scottgonzalez commented on lib/client.js:
```diff
@@ -0,0 +1,89 @@
+
+	fn = fn || function() {};
+
+	var reqOptions = {
```
Do the option creation all at once:

```js
var reqOptions = extend({
    ...
}, options );
```

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+	extend = require( "./extend" ),
+	util = require( "util" );
+
+
```
spacing

kartsims replied:
What is wrong with the spacing ?..

scottgonzalez replied:
If you look at the [existing code](https://github.com/scottgonzalez/node-browserstack/blob/1246998566c7d577f01b43dc56710ed1f9a198ed/lib/browserstack.js#L5), you'll see there's only one blank line here.

You don't have to worry about all the spacing issues though, I can fix those when I merge.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+}
+util.inherits( ApiBaseClient, BaseClient );
+
+
```
spacing

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+util.inherits( ApiBaseClient, BaseClient );
+
+
+// public api
```
capitalize API

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+
+// public api
+extend( ApiBaseClient.prototype, {
+
```
spacing

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+			path: this.path( "/worker/" + id + "/screenshot.json" )
+		}, fn );
+	}
+
```
spacing

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+
+});
+
+
```
spacing

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+});
+
+
+// internal api
```
capitalize API

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+
+// internal api
+extend( ApiBaseClient.prototype, {
+
```
spacing

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+
+	server: {
+		host: "api.browserstack.com"
+	},
```
This can't be in the prototype. Properties containing objects are shared across all instances. Please revert to setting this in the constructor.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+
+	updateLatest: function( browsers ) {
+		var latest = this.latest = {},
+		getBrowserId = this.getBrowserId,
```
Bad indentation.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+		this.latestUpdate = new Date();
+		browsers.forEach(function( browser ) {
+			var version = browser[ versionField ],
+			browserId = getBrowserId( browser );
```
Bad indentation.

scottgonzalez commented on lib/api.js:
```diff
@@ -0,0 +1,284 @@
+	}
+});
+
+
```
spacing

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,55 @@
+	extend = require( "./extend" ),
+	util = require( "util" );
+
+
```
spacing

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,55 @@
+}
+util.inherits( ScreenshotClient, BaseClient );
+
+
```
spacing

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,55 @@
+
+// public API
+extend( ScreenshotClient.prototype, {
+
```
spacing

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,55 @@
+			path: this.path( "/" + id + ".json" )
+		}, fn );  
+	}
+
```
spacing

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,55 @@
+
+});
+
+
```
spacing

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,55 @@
+
+// internal API
+extend( ScreenshotClient.prototype, {
+
```
spacing

scottgonzalez commented on lib/screenshot.js:
```diff
@@ -0,0 +1,55 @@
+// internal API
+extend( ScreenshotClient.prototype, {
+
+	server: {
```
See note in API client about `server`.

kartsims commented:
I did the changes, except for spacing issues

I hope the server config thing is done right now, I am not 100% sure of what you meant but I think these new commits should be OK.

scottgonzalez commented:
> I hope the server config thing is done right now, I am not 100% sure of what you meant but I think these new commits should be OK.

I put this together to show you what I was talking about: http://jsbin.com/medulawavi/1/edit?js,output

I'll try to do another review tonight.

scottgonzalez commented on readme.md:
```diff
@@ -152,6 +182,30 @@ Gets the latest version of all browsers.
+
+#### screenshotClient.getScreenshot( options, callback )
+
+Runs a job to take screenshots. Options is an object with the following keys :
```
Looks like you never finished documenting this.

Also missing docs for `getJob()`.

scottgonzalez replied:
I've fixed this locally.

scottgonzalez replied:
Actually, the value passed to the callback isn't correct either. I'll fix that as well.

scottgonzalez commented on readme.md:
```diff
@@ -49,7 +62,24 @@ Worker objects are extended [browser objects](#browser-objects) which contain th
+
+A "screenshot object" is a configuration object to run a screenshot job.
+
+It is a plain object with properties defined on [BrowserStack's Screenshot API reference](https://www.browserstack.com/screenshots/api#generate-screenshots).
```
The BrowserStack docs for this are actually really bad. I'm going to inline the properties.

scottgonzalez commented:
I fixed the documentation and published as v1.3.0.

:star2: Thanks so much for implementing this and sticking through all the updates! :star2:

kartsims commented:
Hurray! This is my first "real" contribution to an open source project :balloon:

I am hooked into Node and JS already... Thank you  for being part of it, for the advice and help in doing things right :+1:

scottgonzalez commented:
You're welcome :-) I've been managing open source projects for quite a while and I can say that for someone new, you did a fantastic job. From [asking for more details before sending a PR](https://github.com/scottgonzalez/node-browserstack/issues/25#issuecomment-154404111) to [designing the new API](https://github.com/scottgonzalez/node-browserstack/issues/25#issuecomment-154520847) and [comparing pros and cons of competing modules](https://github.com/scottgonzalez/node-browserstack/issues/25#issuecomment-154528253) all the way to implementation, you followed a great process for getting engaged and seeing the feature the whole way through.

It was a pleasure working with you. Let me know if you ever have any questions about any of this stuff and I'd be happy to help out.

kartsims commented:
:blush: Thanks for the great feedback and the offer !
